ext/strings.go

```diff
@@ -485,6 +545,293 @@ func join(strs []string) (string, error) {
+		if useUpper {
+			fmtStr = "%X"
+		}
+		switch arg.Type() {
+		case types.StringType, types.BytesType:
+			if arg.Type() == types.BytesType {
+				return fmt.Sprintf(fmtStr, arg.Value().([]byte)), nil
+			} else {
+				return fmt.Sprintf(fmtStr, arg.Value().(string)), nil
+			}
+		case types.IntType:
+			argInt, ok := arg.Value().(int64)
+			if !ok {
+				return "", fmt.Errorf("could not convert \"%s\" to int64", arg.Value())
+			}
+			return fmt.Sprintf(fmtStr, argInt), nil
+		default:
+			return "", fmt.Errorf("only integers, byte buffers, and strings can be formatted as hex, was given %s", arg.Type().TypeName())
+		}
+	}, nil
+}
+
+func formatOctal(lastStrIndex int, formatStr string) (int, clauseImpl, error) {
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		if arg.Type() != types.IntType {
+			return "", fmt.Errorf("octal clause can only be used on integers, was given %s", arg.Type().TypeName())
+		}
+		argIntVal := arg.ConvertToType(types.IntType)
+		argInt, ok := argIntVal.Value().(int64)
+		if !ok {
+			return "", fmt.Errorf("could not convert \"%s\" to int64", argIntVal.Value())
+		}
+		return fmt.Sprintf("%o", argInt), nil
+	}, nil
+}
+
+func parseFormattingClause(lastStrIndex int, formatStr string) (int, clauseImpl, error) {
+	i, precision, err := parsePrecision(lastStrIndex+1, formatStr)
+	if err != nil {
+		return -1, nil, fmt.Errorf("error while parsing precision: %w", err)
+	}
+	r := rune(formatStr[i])
+	switch r {
+	case 's':
+		return formatString(i, formatStr)
+	case 'd':
+		return formatDecimal(i, formatStr)
+	case 'f':
+		return formatFixed(i, formatStr, precision)
+	case 'e', 'E':
+		return formatScientific(i, formatStr, unicode.IsUpper(r), precision)
+	case 'b':
+		return formatBinary(i, formatStr)
+	case 'h', 'H':
+		return formatHex(i, formatStr, unicode.IsUpper(r))
+	case 'o':
+		return formatOctal(i, formatStr)
+	default:
+		return -1, nil, fmt.Errorf("unrecognized formatting clause \"%c\"", r)
+	}
+}
+
+func parsePrecision(lastStrIndex int, formatStr string) (int, *int, error) {
+	i := lastStrIndex
+	if formatStr[i] != '.' {
+		return i, nil, nil
+	}
+	i++
+	var buffer strings.Builder
+	for true {
+		if i >= len(formatStr) {
+			return -1, nil, errors.New("could not find end of precision specifier")
+		} else if !isASCIIDigit(rune(formatStr[i])) {
+			break
+		}
+		buffer.WriteByte(formatStr[i])
+		i++
+	}
+	precision, err := strconv.Atoi(buffer.String())
+	if err != nil {
+		return -1, nil, fmt.Errorf("error while converting precision to integer: %w", err)
+	}
+	return i, &precision, nil
+}
+
+func isASCIIDigit(r rune) bool {
+	return r <= unicode.MaxASCII && unicode.IsDigit(r)
+}
+
+func matchLanguage(locale string) language.Tag {
+	tag, _ := language.MatchStrings(makeMatcher(locale), locale)
+	return tag
+}
+
+func makeMatcher(locale string) language.Matcher {
+	tags := make([]language.Tag, 0)
+	tag, err := language.Parse(locale)
+	// silenly fall back if the locale tag is invalid
```

Comment:
If the developer mistypes a tag the result is the default tag instead of an error? Is this what we want?

Reply:
Fixed - I've changed it so that the tag is checked during env setup.